hibernate-core/src/main/java/org/hibernate/type/ComponentType.java

```diff
@@ -79,14 +80,48 @@ public ComponentType(Component component, int[] originalPropertyOrder, MetadataB
 	}
 
 	public ComponentType(Component component, int[] originalPropertyOrder) {
-		this( component, originalPropertyOrder,
-				component.getComponentClassName() != null
-						&& !isRecord( component.getComponentClass() ) );
+		this( component, originalPropertyOrder, (Boolean) null);
 	}
 
 	public ComponentType(Component component, int[] originalPropertyOrder, boolean mutable) {
-		this.componentClass = component.isDynamic() ? Map.class : component.getComponentClass();
-		this.mutable = mutable;
+		this(component, originalPropertyOrder, (Boolean) mutable);
+	}
+	
+	private ComponentType(final Component component, int[] originalPropertyOrder, Boolean mutable) {
+		// As mutable and componentClass are based on the actual component
+		// type, both variables must be lazily initialized in some cases.
+		// E.g. on code generation by hibernate-tools, the component type
+		// might not yet exist (compound keys), resulting in a
+		// ClassNotFoundException.
```

Comment:
I'm not sure I fully understand this use-case: why would the component class, i.e. the class of the generated composite key, not exist while the `componentClassName` does not return null? Maybe the problem is we shouldn't return a class name at all, or we should resolve the class eagerly when generating the `ComponentType`.

Could you please include a test demonstrating the issue within your PR? This can help us understand the problem better and ensure no regressions will be introduced in the future.

Reply:
The classes are also used by the hibernate-tool project for the code generation. So the to-be-generated XXX**Id** class is present in the DB metadata (because to a compound key) and therefore in the class model but does not yet physically exist.

There is also an active hibernate-tools issue for this: [HBX-2840](https://hibernate.atlassian.net/browse/HBX-2840)

But I will try to create a standalone reproducer.

Reply:
> There is also an active hibernate-tools issue for this: [HBX-2840](https://hibernate.atlassian.net/browse/HBX-2840)

Great, hopefully Koen will be able to take a look soon and if we can identify this as a tools issue I'm sure it will be fixed. If you want, you can also try taking a look and proposing a fix, if possible.

Let us know about the test. If you don't mean to continue contributing through this PR, please close it so we can progress with cleanup in our backlog.

Reply:
I doubt it is a pure TOOLS issue, as the second-pass processing is setup and run by ORM, based on the metadata feed via TOOLS.  See [this line in mapping.BasicValue](https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/mapping/BasicValue.java#L129).
Either TOOLS is not allowed to use the classes from CORE anymore - which would be a huge pain in the a.. (i think) or this use case needs to be handled by ORM.

This commit is the root cause of this issue: https://github.com/hibernate/hibernate-orm/commit/3e8056de8eedfc1ede1114d6f6fea91c91232852


Reproducer is attaches as a separat comment.